libsplinter/src/matrix.rs

```diff
@@ -14,9 +14,35 @@
 
 use std::error::Error;
 use std::fmt;
+use std::time::Duration;
 
 use crate::transport::Connection;
 
+/// Wrapper around payload to include connection id
+#[derive(Debug, Default, PartialEq)]
+pub struct Envelope {
+    id: String,
+    payload: Vec<u8>,
+}
+
+impl Envelope {
+    pub fn new(id: String, payload: Vec<u8>) -> Self {
+        Envelope { id, payload }
+    }
+
+    pub fn id(&self) -> &str {
+        &self.id
+    }
+
+    pub fn payload(&self) -> &[u8] {
+        &self.payload
+    }
+
+    pub fn take_payload(self) -> Vec<u8> {
```

Comment:
It it never cloned (at least in the way mesh uses it) Mesh will take the payload when it tries to send it over the channel. If this fails the payload may be returned by the error raised (on full or disconnected) and then rewrapped in an envelope that will be returned with the error.

Reply:
The other places that we use take with an option has more to do with how the struct can continue to be used, but only can provide one of the items that are being taken.  That is, the struct is not consumed by the take_x, only x is (in effect).

In this case, we want to take apart the envelope and transform it into the internal mesh ide and payload, so we can consume it.  Given that this is fairly low-level struct, I would by fine with making the fields pub and doing way with these methods.  That way, the envelope can be taken apart through mechanisms such as destructuring.

Reply:
The goal of this addition currently was to just move the Envelop used by mesh so it would be available to the Matrix trait. I think we should create a separate story for changing this when we can change the way the Envelope is used everywhere.

Reply:
I missed that it was consuming self. That's not how I understood "take", but I guess maybe it has more than one definition. From is a cleaner way to implement the same thing.

Reply:
into_payload() would not have been confusing either.